pkg/git/git.go

```diff
@@ -0,0 +1,428 @@
+//
+// Copyright 2023 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	GitHubHost    string = "github.com"
+	RawGitHubHost string = "raw.githubusercontent.com"
+	GitLabHost    string = "gitlab.com"
+	BitbucketHost string = "bitbucket.org"
+)
+
+type IGitUrl interface {
+	GitRawFileAPI() string
+	SetToken(token string, httpTimeout *int) error
+	IsPublic(httpTimeout *int) bool
+	DownloadGitRepoResources(url string, destDir string, httpTimeout *int, token string) error
+}
+
+type Url struct {
+	Protocol string // URL scheme
+	Host     string // URL domain name
+	Owner    string // name of the repo owner
+	Repo     string // name of the repo
+	Branch   string // branch name
```

Comment:
only branch is supported? can we do revision instead? I think tag/commitID also need to be supported.

Reply:
Yeah I wasn't sure what the best way to name this was so I just chose `Branch`, but it does support them all:

https://github.com/devfile/library/tree/0ce592a416fb185564516353891a45016ac7f671
https://github.com/devfile/library/blob/0ce592a416fb185564516353891a45016ac7f671/devfile.yaml

- 	`branch = 0ce592a416fb185564516353891a45016ac7f671`

https://github.com/devfile/library/tree/main
https://github.com/devfile/library/blob/main/devfile.yaml

- 	`branch = main`

https://github.com/devfile/library/tree/v2.2.0
https://github.com/devfile/library/blob/v2.2.0/devfile.yaml

- 	`branch = v2.2.0`